[V1] Adding min tokens/repetition/presence/frequence penalties to V1 sampler

sroy745 opened the pull request:
In this PR we add the logic to support min_tokens, repetition penalty, presence penalty, frequency penalty aware sampling for V1. We make the following changes in the pr

1. Add logic in v1/sample/sampler.py and v1/sample/metadata.py to handle the new sampling parameters.
2. Added tests in tests/v1/sample/test_sampler.py to test out the new functionality. Also added a test in tests/v1/engine/test_engine_core.py to have a basic e2e test with some of these sampling params set
3.  Moved some of the common penalty logic from vllm/model_executor/layers/sampler.py to vllm/utils.py and reused it in the v1 code.
4. In the penalizing code changed the logic for repetition penalty to change the logits inplace. In the current code we might be creating a copy of the logits when applying that logic. Hence modified it to apply the changes in place since we are doing the same for the other penalties.


Note that 4) is not fully related to this PR.

sroy745 committed:
Merge pull request #1 from vllm-project/main

Pull from head

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Add options for min_tokens/repetition etc penalties to V1 sampler

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Fixes

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Add tests

sroy745 committed:
Add tests

sroy745 committed:
Fix format

sroy745 committed:
Comments

sroy745 committed:
Tests

sroy745 committed:
Fixes

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 commented:
@WoosukKwon / @afeldman-nm the pr is ready for review. PTAL when you get a chance.

sroy745 committed:
Fix tests

sroy745 committed:
Fixes

Signed-off-by: Sourashis Roy <sroy@roblox.com>

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -820,6 +822,28 @@ def make_sampling_metadata(
             self.top_k[:self.num_reqs].copy_(
                 self.top_k_cpu_tensor[:self.num_reqs], non_blocking=True)
+
+        output_token_ids: List[List[int]] = []
+        prompt_token_ids: List[List[int]] = []
+        frequency_penalties: List[float] = []
+        presence_penalties: List[float] = []
+        repetition_penalties: List[float] = []
+        min_tokens: List[int] = []
+        stop_token_ids: List[set[int]] = []
```
QQ: Do we have to use this python representations? Now that the worker is stateful with the "persistent batch", can we keep these in tensors?

sroy745 replied:
**Tensor representation**
I think we can convert frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids into tensors because they won't change in every step. 

**Non Tensor representation**
- output_token_ids will be updated in every step and so we might need to maintain it as a list. 
- min_tokens and stop_token_ids are not used directly. They are used in combination with output_token_ids to determine the the indices of the logits that need to be set to -inf.

Do you think it would be ok to convert   frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids  into tensor representation and retain the rest (output_token_ids, min_tokens, stop_token_ids) as python representations?

WoosukKwon replied:
@sroy745 Thanks for the explanation. Sounds good to me!

WoosukKwon replied:
In the future, we could do more optimizations based on the states. For example, we can actually cache and update the statistics (histogram), so that we don't need to recompute it every step.

sroy745 replied:
@WoosukKwon as discussed moved frequency_penalties, presence_penalties, repetition_penalties, prompt_token_ids into tensor representations. I will address the incremental update of output_token statistics in a separate pr (I have added a TODO for it). PTAL.

I am working on adding some tests for sampling metadata management in InputBatch.

sroy745 committed:
Fixes

sroy745 committed:
Fixes

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge remote-tracking branch 'origin/main' into sroy-v1-sampling

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -832,6 +900,44 @@ def make_sampling_metadata(
+                        requests, self.vocab_size, device=self.device)
+
+        output_token_ids: List[List[int]] = []
+        min_tokens: List[int] = []
+        stop_token_ids: List[set[int]] = []
```
Why don't we cache these values in `InputBatch` rather than recreating them every time?

sroy745 replied:
Done

WoosukKwon commented on vllm/v1/sample/sampler.py:
```diff
@@ -156,3 +167,64 @@ def _apply_top_k_top_p(
+        logits[tuple(zip(*min_tokens_logits_to_penalize))] = -float("inf")
+
+
+def _apply_penalties(logits: torch.Tensor, prompt_token_ids: torch.Tensor,
```
Can we share the method with V0 sampler? It seems the code is almost duplicated.

sroy745 replied:
Moved the common to apply_penalties in vllm/model_executor/layers/utils.py

WoosukKwon commented on vllm/v1/sample/sampler.py:
```diff
@@ -156,3 +167,64 @@ def _apply_top_k_top_p(
+    """
+    min_tokens_logits_to_penalize: List[Tuple[int, int]] = []
+    for index, min_token in enumerate(min_tokens):
+        if (min_token > 0 and len(output_token_ids[index]) < min_token):
```
Why do we check `min_token` > 0?

sroy745 replied:
Removed.

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -697,6 +701,44 @@ def __init__(
+                                               dtype=torch.float,
+                                               device=device)
+        self.frequency_penalties_cpu_tensor = torch.empty(
+            (max_num_reqs, vocab_size),
```
Dumb question: Why is it `(max_num_reqs, vocab_size)` instead of just `(max_num_reqs, 1)`? Can we keep this tensor small and use broadcasting for actual computation?

sroy745 replied:
Done

WoosukKwon reviewed (changes requested):
@sroy745 Thanks for doing this! While I'm excited that now V1 can support those sampling params, I think this PR can be a bit polished more to make them better fit into the new persistent-batch-style input prep. Please check my comments.

sroy745 committed:
Merge branch 'vllm-project:main' into main

sroy745 committed:
Merge remote-tracking branch 'origin/main' into sroy-v1-sampling

sroy745 committed:
Addressing comments

sroy745 committed:
Remove debug prints

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Adding utils

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Fixes

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
More fixes

sroy745 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -832,6 +900,44 @@ def make_sampling_metadata(
+                self.repetition_penalties[:self.num_reqs].copy_(
+                    self.repetition_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                # The prompt tokens are used only for applying penalties during
+                # the sampling process. Hence copy these tensors only when
+                # there are requests which need penalties to be applied.
+                self.prompt_tokens_tensor = \
+                    self._construct_prompt_tokens_tensor(
+                        requests, self.vocab_size, device=self.device)
```
I am constructing this tensor only when skip_copy is False and not at every step. However I do construct this tensor using the prompt_token_id list. One alternative I can think of is to maintain the prompt tensors (on gpu) for each of these requests and pass the list of prompt tensors to the sampler. This will avoid converting each of the prompt token id list to a tensor every time there is a batch compaction. However in this case the sampler might need to construct a tensor from this list and then process it further. Is the second option what you are referring to here?

WoosukKwon replied:
OK I was thinking whether we can retain `prompt_token_ids` (just like other tensors in the input batch) rather than reconstructing it.
However, after second thoughts, I think we can keep the current version since the `prompt_token_ids` tensor can be very large and we will eventually manage other states (e.g., bin counts) instead of `prompt_token_ids` for penalties. I'm OK with the current implementation.

sroy745 commented on vllm/utils.py:
```diff
@@ -1659,3 +1659,20 @@ def resolve_obj_by_qualname(qualname: str) -> Any:
     return getattr(module, obj_name)
+
+
+def get_token_bin_counts_and_mask(
```
Created vllm/model_executor/layers/utils.py and moved this method and apply_penalties there

sroy745 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -842,8 +948,39 @@ def make_sampling_metadata(
 
+    def _construct_prompt_tokens_tensor(
+        self, requests, vocab_size: int,  device: torch.device) \
+        -> torch.Tensor:
```
Done

sroy745 committed:
Format

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Some more tests

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Format

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 commented:
Hi @WoosukKwon , thanks for the reviews. Addressed your comments. Also added the tests for sampling metadata management in test_gpu_model_runner.py. PTAL

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -842,8 +952,42 @@ def make_sampling_metadata(
+            output_token_ids=output_token_ids,
+            min_tokens=self.min_tokens[:self.num_reqs],
+            stop_token_ids=self.stop_token_ids[:self.num_reqs],
+            no_penalties=self.no_penalties
```
```suggestion
            no_penalties=self.no_penalties,
```

sroy745 committed:
Merge branch 'main' into sroy-v1-sampling

sroy745 committed:
Merge

sroy745 committed:
Rename test file

sroy745 committed:
Tests

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Changes

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Comments

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 commented:
Thanks for the review. Addressed your comments.

WoosukKwon committed:
Merge branch 'main' into sroy-v1-sampling

sroy745 committed:
Only pass output token_ids to sampler

sroy745 commented:
Hi @WoosukKwon as discussed modified make_sampling_metadata to take as input output_token_ids instead of requests. PTAL.

sroy745 committed:
Dummy

sroy745 committed:
Rerun tests

sroy745 committed:
Remove tolist for prompts

sroy745 committed:
Add TODO

sroy745 committed:
Dummy

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Dummy

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Rerun tests

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Rerun tests

Signed-off-by: Sourashis Roy <sroy@roblox.com>

WoosukKwon committed:
Merge branch 'main' into sroy-v1-sampling

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
optimize

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

sroy745 commented on vllm/v1/worker/gpu_input_batch.py:
```diff
@@ -241,15 +319,48 @@ def condense(self, empty_req_indices: List[int]) -> None:
+                # The prompt tokens are used only for applying penalties during
+                # the sampling process. Hence copy these tensors only when
+                # there are requests which need penalties to be applied.
+                prompt_token_ids = self._make_prompt_token_ids_tensor()
```
@WoosukKwon wondering if we can make prompt_token_ids into a class variable? We are reusing the already computed tensor if skip_copy is True. So for skip_copy==True we might end up using a None value for this tensor. I think this might be causing tests/v1/engine/test_engine_core.py::test_engine_core_advanced_sampling to fail

WoosukKwon replied:
Nice catch! I didn't consider that case. Thanks for the fix!

sroy745 committed:
Make prompt_token_ids a class variable

Signed-off-by: Sourashis Roy <sroy@roblox.com>

sroy745 committed:
Merge branch 'vllm-project:main' into sroy-v1-sampling

WoosukKwon reviewed (approved):
LGTM. Thanks for the great work!
